src/qt/optionsmodel.cpp

```diff
@@ -273,322 +252,345 @@ int OptionsModel::rowCount(const QModelIndex & parent) const
+// read QSettings values and return them
+QVariant OptionsModel::data(const QModelIndex & index, int role) const
 {
-    const bool prune = prune_target_gb > 0;
-    if (prune) {
-        QSettings settings;
-        settings.setValue("nPruneSize", prune_target_gb);
+    if(role == Qt::EditRole)
+    {
+        return getOption(OptionID(index.row()));
     }
-    SetPruneEnabled(prune, force);
+    return QVariant();
 }
 
-// read QSettings values and return them
-QVariant OptionsModel::data(const QModelIndex & index, int role) const
+// write QSettings values
+bool OptionsModel::setData(const QModelIndex & index, const QVariant & value, int role)
 {
+    bool successful = true; /* set to false on parse error */
     if(role == Qt::EditRole)
     {
-        QSettings settings;
-        switch(index.row())
-        {
-        case StartAtStartup:
-            return GUIUtil::GetStartOnSystemStartup();
-        case ShowTrayIcon:
-            return m_show_tray_icon;
-        case MinimizeToTray:
-            return fMinimizeToTray;
-        case MapPortUPnP:
+        successful = setOption(OptionID(index.row()), value);
+    }
+
+    Q_EMIT dataChanged(index, index);
+
+    return successful;
+}
+
+QVariant OptionsModel::getOption(OptionID option) const
+{
+    auto setting = [&]{ return node().getPersistentSetting(SettingName(option)); };
+
+    QSettings settings;
+    switch (option) {
+    case StartAtStartup:
+        return GUIUtil::GetStartOnSystemStartup();
+    case ShowTrayIcon:
+        return m_show_tray_icon;
+    case MinimizeToTray:
+        return fMinimizeToTray;
+    case MapPortUPnP:
 #ifdef USE_UPNP
-            return settings.value("fUseUPnP");
+        return SettingToBool(setting(), DEFAULT_UPNP);
 #else
-            return false;
+        return false;
 #endif // USE_UPNP
-        case MapPortNatpmp:
+    case MapPortNatpmp:
 #ifdef USE_NATPMP
-            return settings.value("fUseNatpmp");
+        return SettingToBool(setting(), DEFAULT_NATPMP);
 #else
-            return false;
+        return false;
 #endif // USE_NATPMP
-        case MinimizeOnClose:
-            return fMinimizeOnClose;
-
-        // default proxy
-        case ProxyUse:
-            return settings.value("fUseProxy", false);
-        case ProxyIP:
-            return GetProxySetting(settings, "addrProxy").ip;
-        case ProxyPort:
-            return GetProxySetting(settings, "addrProxy").port;
-
-        // separate Tor proxy
-        case ProxyUseTor:
-            return settings.value("fUseSeparateProxyTor", false);
-        case ProxyIPTor:
-            return GetProxySetting(settings, "addrSeparateProxyTor").ip;
-        case ProxyPortTor:
-            return GetProxySetting(settings, "addrSeparateProxyTor").port;
+    case MinimizeOnClose:
+        return fMinimizeOnClose;
+
+    // default proxy
+    case ProxyUse:
+        return ParseProxyString(SettingToString(setting(), "")).is_set;
```

Comment:
This diff is now in commit ce530d566f `Migrate -prune setting from QSettings to settings.json`, but it has to be amended to commit dd3d146f72 `Migrate -proxy and -onion settings from QSettings to settings.json`

```diff
--- a/src/qt/optionsmodel.cpp
+++ b/src/qt/optionsmodel.cpp
@@ -338,21 +364,21 @@ QVariant OptionsModel::getOption(OptionID option) const
 #endif // USE_NATPMP
     case MinimizeOnClose:
         return fMinimizeOnClose;
 
     // default proxy
     case ProxyUse:
-        return ParseProxyString(SettingToString(node().getPersistentSetting("proxy"), "")).is_set;
+        return ParseProxyString(SettingToString(setting(), "")).is_set;
     case ProxyIP:
         return m_proxy_ip;
     case ProxyPort:
         return m_proxy_port;
 
     // separate Tor proxy
     case ProxyUseTor:
-        return ParseProxyString(SettingToString(node().getPersistentSetting("onion"), "")).is_set;
+        return ParseProxyString(SettingToString(setting(), "")).is_set;
     case ProxyIPTor:
         return m_onion_ip;
     case ProxyPortTor:
         return m_onion_port;
 
 #ifdef ENABLE_WALLET
@@ -430,56 +456,56 @@ bool OptionsModel::setOption(OptionID option, const QVariant& value)
         settings.setValue("fMinimizeOnClose", fMinimizeOnClose);
         break;
 
     // default proxy
     case ProxyUse:
         if (changed()) {
-            node().updateSetting("proxy", ProxyString(value.toBool(), m_proxy_ip, m_proxy_port));
+            update(ProxyString(value.toBool(), m_proxy_ip, m_proxy_port));
             setRestartRequired(true);
         }
         break;
     case ProxyIP:
         if (changed()) {
             m_proxy_ip = value.toString();
             if (getOption(ProxyUse).toBool()) {
-                node().updateSetting("proxy", ProxyString(true, m_proxy_ip, m_proxy_port));
+                update(ProxyString(true, m_proxy_ip, m_proxy_port));
                 setRestartRequired(true);
             }
         }
         break;
     case ProxyPort:
         if (changed()) {
             m_proxy_port = value.toString();
             if (getOption(ProxyUse).toBool()) {
-                node().updateSetting("proxy", ProxyString(true, m_proxy_ip, m_proxy_port));
+                update(ProxyString(true, m_proxy_ip, m_proxy_port));
                 setRestartRequired(true);
             }
         }
         break;
 
     // separate Tor proxy
     case ProxyUseTor:
         if (changed()) {
-            node().updateSetting("onion", ProxyString(value.toBool(), m_onion_ip, m_onion_port));
+            update(ProxyString(value.toBool(), m_onion_ip, m_onion_port));
             setRestartRequired(true);
         }
         break;
     case ProxyIPTor:
         if (changed()) {
             m_onion_ip = value.toString();
             if (getOption(ProxyUseTor).toBool()) {
-                node().updateSetting("onion", ProxyString(true, m_onion_ip, m_onion_port));
+                update(ProxyString(true, m_onion_ip, m_onion_port));
                 setRestartRequired(true);
             }
         }
         break;
     case ProxyPortTor:
         if (changed()) {
             m_onion_port = value.toString();
             if (getOption(ProxyUseTor).toBool()) {
-                node().updateSetting("onion", ProxyString(true, m_onion_ip, m_onion_port));
+                update(ProxyString(true, m_onion_ip, m_onion_port));
                 setRestartRequired(true);
             }
         }
         break;
 
 #ifdef ENABLE_WALLET
```